torch/csrc/distributed/c10d/NCCLUtils.cpp

```diff
@@ -80,14 +64,30 @@ std::shared_ptr<NCCLComm> NCCLComm::split(
   auto comm = std::make_shared<NCCLComm>();
   // This call will block until the source communicator is initialized
   auto sourceComm = source->getNcclComm();
+#ifndef NCCL_HAS_COMM_NONBLOCKING
   C10D_NCCL_CHECK(
       ncclCommSplit(sourceComm, color_id, rank, &(comm->ncclComm_), &config),
       std::nullopt);
+#else
+  // After calling ncclCommSplit in non-blocking mode, we should wait for the
+  // source communicator to be out of ncclInProgress state.
+  // Reason 1:
+  //   it's unsafe to call new operations on the parent comm while it's in
+  //   ncclInProgress state.
```

Comment:
line 66 "auto sourceComm = source->getNcclComm();" should have already guaranteed that the parent comm is ready?

Reply:
The wait here refers to *AFTER* calling `ncclCommSplit` -- we need to wait on `sourceComm` to make sure it gives a valid pointer to child comm.

Reply:
Still confused, before we call ncclCommSplit here, the parent comm should have already been ncclSuccess state instead of ncclInProgress state, because there is already a wait in getNcclComm(). We can discuss this offline as it seems I misunderstood something

Reply:
Would this discussion help clarify?
https://github.com/NVIDIA/nccl/issues/1472#issuecomment-2402892251

Reply:
Then you might need update the comment? because the parent comm is ready/initialized, but it is the child comm which is not initialized, so we need to wait here.